node/network/availability-distribution/src/lib.rs

```diff
@@ -0,0 +1,1552 @@
+	/// Ignores all non existant relay parents, so this can be used directly with a peers view.
+	/// Returns a map from candidate hash -> receipt
+	fn cached_live_candidates_unioned<'a>(
+		&'a self,
+		relay_parents: impl IntoIterator<Item = &'a H256> + 'a,
+	) -> HashMap<H256, CommittedCandidateReceipt> {
+		let relay_parents_and_ancestors = relay_parents
+			.into_iter()
+			.map(|relay_parent| {
+				self
+					.per_relay_parent
+					.get(relay_parent)
+					.into_iter()
+					.map(|per_relay_parent| {
+						per_relay_parent.ancestors.iter().cloned()
+					})
+					.flatten()
+					.chain(iter::once(relay_parent.clone()))
+			})
+			.flatten()
+			.collect::<HashSet<H256>>();
+
+		relay_parents_and_ancestors
+			.into_iter()
+			.filter_map(|relay_parent_or_ancestor| self.receipts.get(&relay_parent_or_ancestor))
+			.map(|receipt_set| receipt_set.into_iter())
+			.flatten()
+			.map(|receipt| (candidate_hash_of(receipt), receipt.clone()))
+			.collect::<HashMap<H256, CommittedCandidateReceipt>>()
+	}
+
+	async fn add_relay_parent<Context>(
+		&mut self,
+		ctx: &mut Context,
+		relay_parent: H256,
+	) -> Result<()>
+	where
+		Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+	{
+		let candidates =
+			query_live_candidates(ctx, self, std::iter::once(relay_parent.clone())).await?;
+
+		// register the relation of relay_parent to candidate..
+		// ..and the reverse association.
+		for (relay_parent_or_ancestor, receipt) in candidates.clone() {
+			self.reverse.insert(
+				candidate_hash_of(&receipt),
+				relay_parent_or_ancestor.clone(),
+			);
+			self.per_candidate
+				.entry(candidate_hash_of(&receipt))
+				.or_default();
+			self.receipts
+				.entry(relay_parent_or_ancestor)
+				.or_default()
+				.insert(receipt);
+		}
+
+		// collect the ancestors again from the hash map
+		let ancestors = candidates
+			.iter()
+			.filter_map(|(ancestor_or_relay_parent, _receipt)| {
+				if ancestor_or_relay_parent == &relay_parent {
+					None
+				} else {
+					Some(ancestor_or_relay_parent.clone())
+				}
+			})
+			.collect::<HashSet<H256>>();
+
+		// mark all the ancestors as "needed" by this newly added relay parent
+		for ancestor in ancestors.iter() {
+			self.ancestry
+				.entry(ancestor.clone())
+				.or_default()
+				.insert(relay_parent.clone());
+		}
+
+		let per_relay_parent = self
+			.per_relay_parent
+			.get_mut(&relay_parent)
+			.expect("Relay parent is initialized on overseer signal. qed");
+		per_relay_parent.ancestors = ancestors;
+
+		Ok(())
+	}
+
+	fn remove_relay_parent(&mut self, relay_parent: &H256) -> Result<()> {
+		// we might be ancestor of some other relay_parent
+		if let Some(ref mut descendants) = self.ancestry.get_mut(relay_parent) {
+			// if we were the last user, and it is
+			// not explicitly set to be worked on by the overseer
+			if descendants.is_empty() {
+				// remove from the ancestry index
+				self.ancestry.remove(relay_parent);
```

Comment:
I guess we could use the entry API here, but I actually think that as currently written it's cleaner. And the performance effect of the entry API will be negligible.

Reply:
I initially had the entry API here, but that would leave a `ancestry` entries there which would hold a `HashSet`, and I did not see a way to use the `Entry` based `API` without resorting to an explicit `match` statement which would make it more convoluted than the above imo.